parser/parser/cc/grammars/typedruby.ypp

```diff
@@ -1653,6 +1654,21 @@ lbrace_cmd_block_start:
                       $$ = driver.alloc.node_list(driver.build.splat(self, $1, $2));
                     }
 
+                     // While scanning an array, if we hit any of these tokens in places where a terminating `]` would
+                     // fit, we should stop scanning the array and report an error.
+array_premature_end: eof
+                   | kRESCUE
+                   | kENSURE
+                   | kEND
+                   | kTHEN
+                   | kELSIF
+                   | kELSE
+                   | kWHEN
+                   | kIN
+                   | tRPAREN
+                   | tRCURLY
+                   | tCOLON
```

Comment:
This seems a little suspicious. I'm worried that this is going to be brittle / cause confusion when the upstream ruby grammar changes in the future in such a way that makes one of these token somehow expected, or if new tokens are added to the lexer but not added  here.

Reply:
I think there are two potential maintenance headaches here:

1. The grammar changes in the future, such that one of these tokens is now expected. In this case, it would manifest as a shift/reduce conflict which bison should flag for us, and it probably wouldn't take too long to work back from there to the fact that this production is implicated. In that case, we'd have to trim one of the tokens from this rule and it is possible that some recoverable parses would become unrecoverable, or at least not recover as gracefully.
2. The grammar changes in the future, such that new tokens are added that are not expected here. In this case, we wouldn't get any guidance from bison, but I don't think this is such a big deal, since it won't result in incorrect parses of valid syntax---it might just take us a while to realize that there are new tokens that need to be added here to improve the quality of error recovery.

Neither one of these is ideal, but I'm not sure they're showstoppers.